### ISSUE TITLE
incompatibility reading boolean data written using pynwb/h5py

### ISSUE THREAD

**Issue by bendichter** [NeurodataWithoutBorders/matnwb#206]
HDF5 [does not natively have a boolean format](https://support.hdfgroup.org/HDF5/hdf5-quest.html#bool). h5py (and by extension pynwb) handles this by creating an enumerated type with an underlying data type of uint8 that is constrained to 0-1 and maps onto the values "FALSE" and "TRUE" (see [documentation](http://docs.h5py.org/en/stable/faq.html#what-datatypes-are-supported)). h5py knows to write data this way, and it also knows to convert this type of data automatically to a boolean numpy array on read. However, MATLAB's HDF5 API does not use this convention, and reads the data according to the enumerated type mapping defined in the dataset: "TRUE" and "FALSE". The data is stored efficiently in the HDF5 file, but is not read efficiently by MATLAB, and requires the user to wrangle the data back to boolean form. The pipeline is working as expected on the python side, but we have a bug/incompatibility when boolean data written by PyNWB is read using MatNWB.

**Comment by ln-vidrio** [comment]
Related to #77

**Comment by bahanonu** [comment]
@bendichter Thanks for raising this issue. Specifically, MATLAB will read a multi-dimensional/nested cell array of `TRUE` and `FALSE` (e.g. `512x512x88` cell array of strings).

Fiji appears to have issues reading, but ImageJ loads them fine.

Is the way h5py stores 0-1 uint8 any more efficient (space-wise) after compression that storing it as just a uint8 matrix of 0s and 1s?

**Comment by bendichter** [comment]
@bahanonu Thanks for the additional context. I think the only advantage of storing them this way as opposed to 0s and 1s in `uint8` is that h5py knows to convert these values to boolean on read. If h5py reads a `uint8` array, it will keep the values as `uint8`.

**Comment by bahanonu** [comment]
@bendichter Sounds good, makes sense to keep the h5py compatibility in that case. 

Are you thinking of including a small function within matnwb repo to convert to a logical array (as opposed to each user writing their own)?

Tested the issue in `R` with `rhdf5` package and that same dataset loads as expected after coercing into a numeric array. Using ophys data in https://gui.dandiarchive.org/#/file-browser/folder/5e834c3b3c4aab7fa53666ad.

**Comment by bendichter** [comment]
> Are you thinking of including a small function within matnwb repo to convert to a logical array (as opposed to each user writing their own)?

Yes, that is the goal of this ticket :-)

**Comment by bendichter** [comment]
Maybe the best approach is to use [this example code](https://www.mathworks.com/help/matlab/ref/h5t.enum_valueof.html):

```matlab
fid = H5F.open('example.h5');
dset_id = H5D.open(fid,'/g3/enum');
type_id = H5D.get_type(dset_id);
num_members = H5T.get_nmembers(type_id);
for j = 1:num_members
    member_name{j} = H5T.get_member_name(type_id,j-1);
    member_value(j) = H5T.enum_valueof(type_id,member_name{j});
end
```

running that on an h5py-style boolean, I get

```
>> member_name =

  1×2 cell array

    {'FALSE'}    {'TRUE'}

>> member_value

member_value =

  1×2 int8 row vector

   0   1
```

We could use this as a test and then access the data via `H5D.read(dset_id)`, which outputs `uint8` data. Ideally we would then convert that to `boolean`. But of course we also need to support lazy reading, and if doing both of those is tough we can skip the final conversion for now. It's really just cosmetic anyway.

**Comment by lawrence-mbf** [comment]
Fixed via #77
